Refactor CI workflow

gmazzap opened the pull request:
- "Static" QA, which doesn't need to run on multiple PHP versions, is now executed in a separate job, instead of relying on a condition. The idea is in future to use the same job to run a static analizer like Psalm.
- QA checks are also executed on pull requests
- The `"ci skip"` in the commit message is now *natively* supported by GitHub actions. So I'm using two different messages to skip "selectively" the two jobs: `"skip qa"`  to skip static analysis, and  `"skip tests"`  to skip unit tests
- Both jobs (static qa & unit tests) use `cs2pr` to format output.
- An env variable `USE_COVERAGE` is used to determine if coverage should be used or not. That simplify the checks in various steps.
- The two "install dependencies" steps (one ignoring platform reqs, the other not), are merged in one, using a matrix variable to make that distinction
- The condition to migrate configuration for PHPUnit is not anymore based on PHP version and other matrix variables, but instead by checking PHPUnit version "grepping" the output `composer show phpunit/phpunit`: we now migrate configuration if PHPUnit is in version 9.
- The two steps to run unit tests (with and without coverage) are now merged into one, using an "inline" condition in the command (easier now that we have `USE_COVERAGE` env var)
- The update of codecov.io only happens when pushing on master: does not really make sense to update it in other branches of for pull requests

Note: somewhere else I had issues with the "parallel lint" phar installed via the "setup-php" action: sometimes it installs a version that is not compatible with PHP 8.1. That is why here I'm installing the tool via Composer, ensuring latest version is installed.

jrfnl commented on .github/workflows/php-qa.yml:
```diff
@@ -1,77 +1,95 @@
+                uses: ramsey/composer-install@v2
 
+            -   name: Check code styles
+                run: |
+                    ./vendor/bin/phpcs -ps . --standard=PHPCompatibility --ignore=*/vendor/* --extensions=php --basepath=./ --report-full --report-checkstyle="phpcs-report.xml" --runtime-set testVersion 5.6- --runtime-set ignore_warnings_on_exit 1
+                    cs2pr --graceful-warnings phpcs-report.xml
```
Why are you combining two actions in one job here ?

gmazzap replied:
Semantically is one action. I want to run PHPCS and bave both standard output in logs and PR-compliant output for PRs. PHPCS does not have a way to do that in one command so I use two.
IMO it's 2 commands, but its 1 step.

jrfnl commented on .github/workflows/php-qa.yml:
```diff
@@ -1,77 +1,95 @@
-                - php-versions: '8.2'
-                  dependency-versions: 'highest'
+            -   name: Setup PHP
+                uses: shivammathur/setup-php@v2
+                with:
+                    php-version: 7.4
+                    coverage: none
+                    tools: cs2pr
 
-        continue-on-error: ${{ matrix.php-versions == '8.2' }}
+            -   name: Install dependencies
+                uses: ramsey/composer-install@v2
 
+            -   name: Check code styles
+                run: |
+                    ./vendor/bin/phpcs -ps . --standard=PHPCompatibility --ignore=*/vendor/* --extensions=php --basepath=./ --report-full --report-checkstyle="phpcs-report.xml" --runtime-set testVersion 5.6- --runtime-set ignore_warnings_on_exit 1
+                    cs2pr --graceful-warnings phpcs-report.xml
+
+    unit-tests:
+        runs-on: ubuntu-latest
+        if: ${{ !contains(github.event.head_commit.message, 'skip tests') }}
+        env:
+            USE_COVERAGE: 'no'
+        strategy:
+            fail-fast: false
```
Why are you changing `fail-fast` ?

gmazzap replied:
Because when there's failure I wanna see if the problem exists only on that version or it is a "generic" error. With fail-fast all versions fail if one fails, and I don't have the chance to know if it was a version-specifc error.

jrfnl commented on .github/workflows/php-qa.yml:
```diff
@@ -1,77 +1,95 @@
+            -   name: Setup PHP
+                uses: shivammathur/setup-php@v2
+                with:
+                    php-version: 7.4
+                    coverage: none
+                    tools: cs2pr
 
-        continue-on-error: ${{ matrix.php-versions == '8.2' }}
+            -   name: Install dependencies
+                uses: ramsey/composer-install@v2
 
+            -   name: Check code styles
+                run: |
+                    ./vendor/bin/phpcs -ps . --standard=PHPCompatibility --ignore=*/vendor/* --extensions=php --basepath=./ --report-full --report-checkstyle="phpcs-report.xml" --runtime-set testVersion 5.6- --runtime-set ignore_warnings_on_exit 1
+                    cs2pr --graceful-warnings phpcs-report.xml
+
+    unit-tests:
+        runs-on: ubuntu-latest
+        if: ${{ !contains(github.event.head_commit.message, 'skip tests') }}
+        env:
+            USE_COVERAGE: 'no'
+        strategy:
+            fail-fast: false
+            matrix:
+                php-ver: [ '5.6', '7.0', '7.1', '7.2', '7.3', '7.4', '8.0', '8.1' ]
```
Why are you changing the variable name ?

gmazzap replied:
Copy and paste from other projects.
I wanna have identical workflow with the final target of moving to reusable workflows.

jrfnl commented on .github/workflows/php-qa.yml:
```diff
@@ -1,77 +1,95 @@
-              run: composer phpcompat
+            -   name: Lint PHP sources
+                if: ${{ matrix.dependency-versions == 'highest' }}
+                run: ./vendor/bin/parallel-lint ./src/ --checkstyle | cs2pr
```
Why are you no longer linting the `/tests/` directory ? If there is a parse error, the test run will error out, but why not error out early by including the test files in the linting check ?

gmazzap replied:
If there's a parse error, the failure will happen as soon as test case is loaded, so _very early_. That means that linting the tests will fail _few milliseconds_ earlier, but in the great majority of cases, when there're no syntax errors, it takes much more time to lint tests. 1 second * 8 versions * _each time_ the workflow run is *much* more time that I can save failing a little bit earlier _only if a failure is there_ (very rarely).

At work, where we pay for GitHub action time, by removing lint for the test folders we saved a non-negligible amount of money (we have _hundreds_ of repositories actively developed).

jrfnl reviewed (changes requested):
@gmazzap I've a had a good look at this PR and remain with a lot of remarks/questions. I do like the general principle of what I think you're trying to do, but I think it needs additional work.

### General remark:

You are making a lot of different changes in one PR (indentation size, matrix variable name changes, logic changes, splitting jobs).

Making each change in a separate commit - or even a separate PR - would make them 1) easier to review and 2) would allow for each change to be a separate decision point (commit can be cleanly removed if decided against it). 

Everything lumped together in one commit in one PR makes it an "all or nothing" decision...
In my experience, this often leads to PRs which don't get merged as they do too much and while 90% is okay, 10% isn't and is holding up the other 90% from getting merged.

### General principle

> "Static" QA, which doesn't need to run on multiple PHP versions, is now executed in a separate job, instead of relying on a condition. The idea is in future to use the same job to run a static analizer like Psalm.

* Would it be an idea to split things into two different workflows (files) ? One which has static analysis/QA jobs and one which has test jobs against a matrix ?
    As GH Actions doesn't allow for re-running an individual build, re-running the workflow when a test fails, woudl currently also re-run the QA check(s). If this is split off to a separate workflow, only the tests would be re-run.
* If split off to a separate workflow, additional QA tasks to be added in the future can then be added as separate jobs instead of being added to the `static-qa` job.
* What about moving the Parallel Lint task to a separate job or even a separate workflow too ?
* What about making the code coverage build a separate job within the `test` workflow ? That would allow to get rid of nearly all conditions.
    It would also allow for using `needs: unit-tests` for the code coverage job to only run the code coverage when the tests pass.

### Maintainability remarks

#### PHPCompatibility

* There is a `phpcompat` script in the `composer.json` file. You are now no longer using that, so chances of the CLI parameter of both scripts (one in the workflow vs the one in `composer.json`) diverging is high as it is easy to forget to update both when something changes.
    I'd like to strongly suggest to continue using the Composer script in the workflow. You can add extra CLI arguments to that like so: `composer phpcompat -- --report-full --report-checkstyle="phpcs-report.xml"`
* As a next iteration: What about adding a `phpcs.xml.dist` file with the configuration to the repo ?
    And if so: what about maybe checking the code for code style consistency too ? Maybe using PSR12 with some customizations ?

#### Yaml file

* It is more common to use 2-space based indentation in Yaml files like this. Why are you changing it to a 4-space based indentation ?
    Note: The indentation change makes the PR harder to review as it obscures the real changes.
* You also seem to have removed a lot of blank lines (and some inline documentation) from the workflow, which makes the workflow harder to read.
* On that note, we could add a YAML linter job to GH Actions to keep all YAML files consistent in formatting ?


### Other remarks

> QA checks are also executed on pull requests

:+1:

> The `"ci skip"` in the commit message is now natively supported by GitHub actions. So I'm using two different messages to skip "selectively" the two jobs: `"skip qa"` to skip static analysis, and `"skip tests"` to skip unit tests

:+1:

> Both jobs (static qa & unit tests) use cs2pr to format output.

Just checking: am I seeing it correctly that `cs2pr` is not used for the tests, but only for the linting in the unit test job ?

> The two "install dependencies" steps (one ignoring platform reqs, the other not), are merged in one, using a matrix variable to make that distinction

The matrix is now inconsistent as _most_ jobs don't have the `composer-options` key. I've seen problems with inconsistent matrixes elsewhere.
What about adding `composer-options: ['']` to the main matrix to ensure that the key is always available ?

> The condition to migrate configuration for PHPUnit is not anymore based on PHP version and other matrix variables, but instead by checking PHPUnit version "grepping" the output `composer show phpunit/phpunit`: we now migrate configuration if PHPUnit is in version 9.

I don't think this is a stable solution.

First off, the migration isn't needed in the first place yet as PHPUnit 9 will work fine with the "old-style" configuration.
Secondly, you can not blindly check for PHPUnit 9. The config format was only changed in PHPUnit 9.3 and the `--migrate-configuration` command only became available in PHPUnit 9.3 as well.

Also note that the current logic will break as soon as PHPUnit 10 comes out, which is when the migration is actually **_needed_**.

Lastly, wouldn't it be simpler to use the output of `"vendor/bin/phpunit" --version` to get the version number ?


> The two steps to run unit tests (with and without coverage) are now merged into one, using an "inline" condition in the command (easier now that we have `USE_COVERAGE` env var)

See my remark above about possibly splitting the code coverage build off to a separate job.


> The update of codecov.io only happens when pushing on master: does not really make sense to update it in other branches of for pull requests

Actually... this effectively means that you will have no idea whether the code coverage changes for a PR and can't evaluate easily whether any tests added in a PR cover the functional change sufficiently.

I would recommend reverting this change.


> Note: somewhere else I had issues with the "parallel lint" phar installed via the "setup-php" action: sometimes it installs a version that is not compatible with PHP 8.1. That is why here I'm installing the tool via Composer, ensuring latest version is installed.

If installing it via Composer is preferred, why not add it to the projects `composer.json` file in `require-dev` ?
And if so, what about adding a script with the command and its arguments to the `composer.json` file too ? That would make it more straight-forward to ensure people run the command locally and in CI with the same arguments.

Also: I know the issue you're talking about. Let me check, I thought a new version had been released which sorted that out, if not, I'll chase that up.

gmazzap commented:
First of all @jrfnl , thanks a lot for the very detailed and informed feedback.

## General remark:

> You are making a lot of different changes in one PR (indentation size, matrix variable name changes, logic changes, splitting jobs).

I know, I'm sorry. The thing is that I'm trying to standardize action workflows across several projects, so what I did here was to copy the workflow from another project and paste here. I saw there were a lot of changes, but thought that being a single file it could be somehow manageable.

# General principle

> Would it be an idea to split things into two different workflows (files) ? One which has static analysis/QA jobs and one which has test jobs against a matrix ?
As GH Actions doesn't allow for re-running an individual build, re-running the workflow when a test fails, would currently also re-run the QA check(s). If this is split off to a separate workflow, only the tests would be re-run.

If test fails, to rerun the test I would first make a commit. And including "skip qa" in the commit message the static analysis would not be performed.
As I said, I'm trying to standardize workflows across projects, and having a single file for PHP QA makes it easier.
And the benefit now that there's one single static QA step is not worth the trouble.


> If split off to a separate workflow, additional QA tasks to be added in the future can then be added as separate jobs instead of being added to the static-qa job.

If I ever manage to use more static analysis steps I will maybe considering splitting. (MAybe not if I've meen successully migrated to resuable workflows).
But doing it now looks overkill for no benefit and makes sync across projects harder.


> What about making the code coverage build a separate job within the test workflow ? That would allow to get rid of nearly all conditions.

I see value in this, but not convinced is a good idea. 
The matrix will be more complex, because I don't want to run the tests in PHP 7.4 with "highest" dependencies twice (once with coverage, once without).
So I would need to remove 7.4 from the "versions" array and add use an `include` for it, where the dependencies type is fixed to "prefer-lowest".
Besides that, because the job also runs "parallel lint", that step (actually 2 steps, installation + execution) would be duplicated. And with the duplication of checkout, setup PHP, and dependencies install, the "tests+coverage" job would be 90% of duplicated code.
So to remove 3 conditions I made a matrix more complex, and added 30 lines of duplicated code. Not sure it's a good deal.


# Maintainability remarks

> There is a phpcompat script in the composer.json file. You are now no longer using that, so chances of the CLI parameter of both scripts (one in the workflow vs the one in composer.json) diverging is high as it is easy to forget to update both when something changes.

I tend to use composer scripts for running scripts _locally_, while CI use something different. 
And once again, I'm more concerned in having workflows syncronized across projects, hopefully switching at some point to [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows). That's imo what will have a _big_ impact on maintenace.
"Hardcoding" a composer script in the CI workflow means I need to have also in other projects...  and that's not the case.
Finally, the idea is to introduce coding standard and `phpcs.xml.dist` file, which will reduce/remove the need of params so this is less of a concern.


> And if so: what about maybe checking the code for code style consistency too ? Maybe using PSR12 with some customizations ?

As mentioned above, yes, that's the idea. I would like to use [Inpsyde coding standard](https://github.com/inpsyde/php-coding-standards), that includes PHPCompqt, but I need to customize them because I can't use scalar types here due to PHP 5.6 support, and I can't rename all the methods to camelCase or I would break all the tests code out there. So I need some time to dedicate to that.


# Yaml file

> It is more common to use 2-space based indentation in Yaml files like this. Why are you changing it to a 4-space based indentation ?

As I said multiple times, I wanna have the same workflow across projects.
And in all other projects I use 4 spaces for YAML just like for other languages (JSON, PHP, JS, XML).
It is "weird" for me to have 4 indentation spaces in all languages but 2 in YAML.


> Note: The indentation change makes the PR harder to review as it obscures the real changes.

Yes, I know, and once again I'm sorry for this. But as said above, this was a copy and paste from another project.


> You also seem to have removed a lot of blank lines (and some inline documentation) from the workflow, which makes the workflow harder to read.

Yes, sorry but I don't like comments in YAML files. For me, they make the file _harder_ to read (especially in GitHub where the color contrast between comments and YAML code is very low). I also think they add "fatigue" to reading a file, which is not useful when the comment says something that is either obvious or not really relevant for the reader.
Regarding blank lines, I don't think I removed any: https://github.com/Brain-WP/BrainMonkey/blob/ci-workflow-refactoring/.github/workflows/php-qa.yml
Or better, I removed a single blank line in before `matrix.include`(https://github.com/Brain-WP/BrainMonkey/blob/master/.github/workflows/php-qa.yml#L22) which I found confusing considering that `include` is part of the matrix.


Other remarks

> am I seeing it correctly that cs2pr is not used for the tests, but only for the linting in the unit test job?

Correct. Is only used for parallel lint which runs in the same job. Reason: I did not find a simple way to make PHPUnit generate checkstyle compliant output. Do you know any?


> The matrix is now inconsistent as most jobs don't have the composer-options key. I've seen problems with inconsistent matrixes elsewhere.
What about adding composer-options: [''] to the main matrix to ensure that the key is always available ?

What kind of problems have you seen? IIRC when I implemented this in other projects I read the GitHub documentation and found out that a using `include` to add a matrix value which is not defined earlier is what GitHub expects in situations like this. 


> First off, the migration isn't needed in the first place yet as PHPUnit 9 will work fine with the "old-style" configuration.

The migration was also there [before this change](https://github.com/Brain-WP/BrainMonkey/blob/master/.github/workflows/php-qa.yml#L62). And it was based on PHP version and dependencies type... which is "weird". I don't need (want) to migrate the config when PHP is above a certain version and composer dependencies mode is "highest", I need (want) to migrate the config when PHPUnit is above 9.3. _That's_ the condition.


> Secondly, you can not blindly check for PHPUnit 9. The config format was only changed in PHPUnit 9.3 and the --migrate-configuration command only became available in PHPUnit 9.3 as well.

In other projects where I'm using this workflow I required PHPUnit like `^5.7 || ^7.5 || ^9.5` which is enough to ensure compat with all PHP versions from 5.6 to 8.1. And with that requirement there's no way PHPUnit >9.0 <9.3 is installed. So in that case checking for `"versions.+\* 9\."` is enough to ensure 9.3+.

Anyway, even using [current constraints](https://github.com/Brain-WP/BrainMonkey/blob/master/composer.json#L35), in practice, there's no way that PHPUnit >9.0 <9.3 is installed as well. Because with `--prefer-lowest` composer will install _lowest_ v7 for PHP 7.3 and _lowest_ v8 for PHP 7.4+. When not using `--prefer-lowest`, PHPUnit version will be _latest_ v8 for PHP 7.2 and _latest_ v9 for PHP 7.3+.
I think that we can even live with what we have, or change the `composer.json` requirement from `... || ^9.0` to ``... || ^9.5`, that's what is already _de facto_ required anyway.

Out of curiosity: IIRC you're the one who set PHPUnit requirement in `composer.json`. What's the reason for a so complex requirement when `^5.7 || ^7.5 || ^9.5` covers all supported PHP versions? Narrower and stricter requirements are more predictable, are resolved faster by Composer, and cache miss are less frequent.


> Lastly, wouldn't it be simpler to use the output of "vendor/bin/phpunit --version" to get the version number ?

Not sure it would be _simpler_, it's still one command and not that shorter. But I think it is indeed _more correct_. So I'll do it, thanks!


> Actually... this effectively means that you will have no idea whether the code coverage changes for a PR and can't evaluate easily whether any tests added in a PR cover the functional change sufficiently.

This make sense. My idea was to not "expose" a coverage number if the PR / branch is not going to be merged, but now I realize that codecov allows me to see coverage per branch, so it makes sense to always push. Thank you for the sanity check.


> If installing it via Composer is preferred, why not add it to the projects composer.json file in require-dev ?

Because I don't want it to be installed locally. This is a sanity check for CI, mostly targeting PRs. I never feel the need to check syntax locally. I've an IDE :)



> Also: I know the issue you're talking about. Let me check, I thought a new version had been released which sorted that out, if not, I'll chase that up.

I know that you know :) While researching the issue I found your comments about it. However, I can tell you I had the issue just this morning in other projects. Even when explicitly declaring in setup-php the latest parallel lint version.

gmazzap commented on .github/workflows/php-qa.yml:
```diff
@@ -1,77 +1,95 @@
+    push:
+    pull_request:
+    workflow_dispatch:
 concurrency:
```
See general comment.
